docs/docs/docs/01-core/data-transfer/01-engine.md

```diff
@@ -0,0 +1,304 @@
+
+##### Handling Schema differences
+
+When a schema diff is discovered with a given schemaStrategy, an error is throw. However, before throwing the error the engine checks to see if there are any schema diff handlers set via `engine.onSchemaDiff(handler)` which allows errors to be bypassed (for example, by prompting the user if they wish to proceed).
+
+A diff handler is an optionally asynchronous middleware function that accepts a `context` and a `next` parameter.
+
+`context` is an object of type `SchemaDiffHandlerContext`
+
+```typescript
+// type Diff can be found in /packages/core/data-transfer/src/utils/json.ts
+type SchemaDiffHandlerContext = {
+  ignoredDiffs: Record<string, Diff[]>;
+  diffs: Record<string, Diff[]>;
+  source: ISourceProvider;
+  destination: IDestinationProvider;
+};
+```
+
+`next` is a function that is called, passing the modified `context` object, to proceed to the next middleware function.
+
+```typescript
+const diffHandler = async (context, next) => {
+  const ignoreThese = {};
+  // loop through the diffs
+  Object.entries(context.diffs).forEach(([uid, diffs]) => {
+    for (const [i, diff] of diffs) {
+      // get the path of the diff in the schema
+      const path = [uid].concat(diff.path).join('.');
+
+      // Allow a diff on the country schema displayName
+      if (path === 'api::country.country.info.displayName') {
+        if (!isArray(context.ignoredDiffs[uid])) {
+          context.ignoredDiffs[uid] = [];
+        }
+        context.ignoredDiffs[uid][i] = diff;
+      }
+    }
+  });
+
+  return next(context);
+};
+
+engine.onSchemaDiff(diffHandler);
+```
+
+After all the schemaDiffHandler middleware has been run, another diff is run between `context.ignoredDiffs` and `context.diffs` and any remaining diffs that have not been ignored are thrown as fatal errors and the engine will abort the transfer.
+
+### Progress Tracking events
+
+The transfer engine allows tracking the progress of your transfer either directly with the engine.progress.data object, or with listeners using the engine.progress.stream PassThrough stream. The engine.progress.data object definition is of type TransferProgress.
+
+Here is an example that logs a message at the beginning and end of each stage, as well as a message after each item has been transferred
+
+```typescript
+const progress = engine.progress.stream;
+
+progress.on(`stage::start`, ({ stage, data }) => {
+  console.log(`${stage} has started at ${data[stage].startTime}`);
+});
+
+progress.on('stage::finish', ({ stage, data }) => {
+  console.log(`${stage} has finished at ${data[stage].endTime}`);
+});
+
+progress.on('stage::progress', ({ stage, data }) => {
+  console.log('Transferred ${data[stage].bytes} bytes / ${data[stage].count} entities');
+});
+```
+
+Note: There is currently no way for a source provider to give a "total" number of records expected to be transferred, but it is expected in a future update.
+
+The following events are available:
```

Comment:
Should we detail the different payloads for each kind of event (stage vs transfer events)?

Reply:
I think that would be a nice iteration but I felt it was too much for this version